In [404]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# from utlis import rm_units_and_conv_int
import utlis

import importlib
importlib.reload(utlis)

<module 'utlis' from 'd:\\1. Projects\\DNA\\Portfolio\\car_market_analysis\\utlis.py'>

### **Merging all datasets into one**

Make sure to uncomment cells in this section in order to create a dataset that contains all data from data scraping.

In [405]:
# from os import listdir

# PATH = 'data_scraping/data/datasets/'

# # Concatenating all datasets
# for i, ds in enumerate(listdir(PATH)):
#     if i == 0:
#         data = pd.read_csv(f'{PATH}{ds}')
#     else:
#         data = pd.concat([data, pd.read_csv(f'{PATH}{ds}')])

# print(data.shape)
# data.head()

In [406]:
# # Translating column names from polish to english
# colnames = {
#     'Marka pojazdu': 'brand',
#     'Model pojazdu': 'model',
#     'Rok produkcji': 'year',
#     'Przebieg': 'mileage',
#     'Pojemność skokowa': 'engine_capacity',
#     'Rodzaj paliwa': 'fuel_type',
#     'Moc': 'power',
#     'Skrzynia biegów': 'transmission',
#     'Napęd': 'drive',
#     'Spalanie W Mieście': 'city_fuel_consumption',
#     'Typ nadwozia': 'body_type',
#     'Liczba drzwi': 'doors',
#     'Liczba miejsc': 'seats',
#     'Kolor': 'color',
#     'Kraj pochodzenia': 'origin_country',
#     'Bezwypadkowy': 'accident_free',
#     'Serwisowany w ASO': 'serviced_at_ASO',
#     'Stan': 'condition',
#     'Cena': 'price'
# }
# data.rename(columns=colnames, inplace=True)

# data.head()

In [407]:
# # Save all data in one file
# data.to_csv('data_scraping/data/all_data.csv', index=False)

# **Data cleaning**

In [408]:
data = pd.read_csv('data_scraping/data/all_data.csv')
data.head()

brand model    year     mileage engine_capacity fuel_type   power  \
0  Audi   SQ7  2017.0  142 800 km       3 956 cm3    Diesel  435 KM   
1  Audi    A3  2010.0  236 000 km       1 395 cm3   Benzyna  125 KM   
2  Audi    A4  2018.0   84 500 km       1 968 cm3    Diesel  190 KM   
3  Audi    Q5  2018.0   38 000 km       1 984 cm3   Benzyna  252 KM   
4  Audi    A7  2016.0   74 885 km       2 995 cm3   Benzyna  333 KM   

   transmission             drive city_fuel_consumption body_type  doors  \
0  Automatyczna       4x4 (stały)             8 l/100km       SUV    5.0   
1      Manualna  Na przednie koła                   NaN   Kompakt    5.0   
2  Automatyczna       4x4 (stały)             5 l/100km     Kombi    5.0   
3  Automatyczna       4x4 (stały)             8 l/100km       SUV    5.0   
4  Automatyczna       4x4 (stały)            10 l/100km   Kompakt    5.0   

   seats   color origin_country accident_free serviced_at_ASO condition  \
0    7.0   Szary         Polska           Tak             Tak   Używane   
1    5.0  Czarny            NaN           Tak             Tak   Używane   
2    5.0   Szary            NaN           NaN             NaN   Używane   
3    5.0   Szary            NaN           NaN             NaN   Używane   
4    5.0   Biały            NaN           NaN             NaN   Używane   

                price  
0  269 000        PLN  
1   29 900        PLN  
2   60 000        PLN  
3   75 000        PLN  
4   75 000        PLN

In [409]:
data.shape

(134057, 19)

This dataset comprises a vast collection of 134,057 observations specifically centered around cars. With 18 informative columns, it offers a diverse range of details pertaining to each vehicle. These encompass essential information such as the brand, model, production year, mileage, and fuel type. 

## **Missing values**


In [410]:
# How many missing values are there in each column
na_count = data.isna().sum()
na_precentage = (na_count / data.shape[0] * 100).round(2)

na_df = pd.DataFrame({'na_count': na_count, 'na_precentage': na_precentage})
na_df.sort_values(by='na_precentage', ascending=False)

na_count  na_precentage
serviced_at_ASO           73520          54.84
accident_free             58608          43.72
city_fuel_consumption     54864          40.93
origin_country            45703          34.09
drive                     17584          13.12
seats                      8409           6.27
engine_capacity            2441           1.82
mileage                    1093           0.82
doors                       735           0.55
power                        83           0.06
transmission                 62           0.05
color                         1           0.00
condition                     0           0.00
brand                         1           0.00
body_type                     1           0.00
model                         1           0.00
fuel_type                     1           0.00
year                          1           0.00
price                         0           0.00

In this dataset, a notable observation is that the majority of columns contain missing values. However, an encouraging aspect is that the columns with missing values appear to be of lesser significance (such as whether the car was serviced at ASO, its accident-free status, and city fuel consumption). On the other hand, the essential data elements, including price, model, brand, and production year, exhibit minimal or no missing values at all. This characteristic suggests that the crucial information crucial for analysis is largely intact, allowing for comprehensive examination and insights into the dataset.

In [411]:
# There is one row with majority of values missing, so I filtered it out. It may be caused by some error during scraping
data[data['brand'].isna()]

brand model  year mileage engine_capacity fuel_type power transmission  \
66878   NaN   NaN   NaN     NaN             NaN       NaN   NaN          NaN   

      drive city_fuel_consumption body_type  doors  seats color  \
66878   NaN                   NaN       NaN    NaN    NaN   NaN   

      origin_country accident_free serviced_at_ASO condition  \
66878            NaN           NaN             NaN   Używane   

                   price  
66878  55 900        PLN

In [412]:
data = data[data['brand'].notna()]

## **Data types**

In [413]:
data.head()

brand model    year     mileage engine_capacity fuel_type   power  \
0  Audi   SQ7  2017.0  142 800 km       3 956 cm3    Diesel  435 KM   
1  Audi    A3  2010.0  236 000 km       1 395 cm3   Benzyna  125 KM   
2  Audi    A4  2018.0   84 500 km       1 968 cm3    Diesel  190 KM   
3  Audi    Q5  2018.0   38 000 km       1 984 cm3   Benzyna  252 KM   
4  Audi    A7  2016.0   74 885 km       2 995 cm3   Benzyna  333 KM   

   transmission             drive city_fuel_consumption body_type  doors  \
0  Automatyczna       4x4 (stały)             8 l/100km       SUV    5.0   
1      Manualna  Na przednie koła                   NaN   Kompakt    5.0   
2  Automatyczna       4x4 (stały)             5 l/100km     Kombi    5.0   
3  Automatyczna       4x4 (stały)             8 l/100km       SUV    5.0   
4  Automatyczna       4x4 (stały)            10 l/100km   Kompakt    5.0   

   seats   color origin_country accident_free serviced_at_ASO condition  \
0    7.0   Szary         Polska           Tak             Tak   Używane   
1    5.0  Czarny            NaN           Tak             Tak   Używane   
2    5.0   Szary            NaN           NaN             NaN   Używane   
3    5.0   Szary            NaN           NaN             NaN   Używane   
4    5.0   Biały            NaN           NaN             NaN   Używane   

                price  
0  269 000        PLN  
1   29 900        PLN  
2   60 000        PLN  
3   75 000        PLN  
4   75 000        PLN

In [414]:
data.dtypes

brand                     object
model                     object
year                     float64
mileage                   object
engine_capacity           object
fuel_type                 object
power                     object
transmission              object
drive                     object
city_fuel_consumption     object
body_type                 object
doors                    float64
seats                    float64
color                     object
origin_country            object
accident_free             object
serviced_at_ASO           object
condition                 object
price                     object
dtype: object

The dataset contains a few data type errors that need to be addressed.  
Because there are missing values in the dataset I convert all numeric columns to float.
* **mileage** should be `float`
* **engine_capacity** should be `float`
* **power** should be `float`
* **city_fuel_consumption** should be `float`
* **price** should be `float`

### **mileage** 

In [415]:
data['mileage'].head()

0    142 800 km
1    236 000 km
2     84 500 km
3     38 000 km
4     74 885 km
Name: mileage, dtype: object

In [416]:
data['mileage'] = utlis.rm_units_and_conv_dtype(data['mileage'], 'km', 'float')

data['mileage'].head()

0    142800.0
1    236000.0
2     84500.0
3     38000.0
4     74885.0
Name: mileage, dtype: float64

### **engine_capacity** 

In [417]:
data['engine_capacity'].head()

0    3 956 cm3
1    1 395 cm3
2    1 968 cm3
3    1 984 cm3
4    2 995 cm3
Name: engine_capacity, dtype: object

In [418]:
data['engine_capacity'] = utlis.rm_units_and_conv_dtype(data['engine_capacity'], 'cm3', 'float')

data['engine_capacity'].head()

0    3956.0
1    1395.0
2    1968.0
3    1984.0
4    2995.0
Name: engine_capacity, dtype: float64

### **power** 

In [419]:
data['power'].head()

0    435 KM
1    125 KM
2    190 KM
3    252 KM
4    333 KM
Name: power, dtype: object

In [420]:
data['power'] = utlis.rm_units_and_conv_dtype(data['power'], 'KM', 'float')

data['power'].head()

0    435.0
1    125.0
2    190.0
3    252.0
4    333.0
Name: power, dtype: float64

### **city_fuel_consumption** 

In [421]:
data['city_fuel_consumption'].head()

0     8 l/100km
1           NaN
2     5 l/100km
3     8 l/100km
4    10 l/100km
Name: city_fuel_consumption, dtype: object

In [422]:
data['city_fuel_consumption'] = utlis.rm_units_and_conv_dtype(data['city_fuel_consumption'], 'l/100km', 'float')

data['city_fuel_consumption'].head()

0     8.0
1     NaN
2     5.0
3     8.0
4    10.0
Name: city_fuel_consumption, dtype: float64

### **price** 

In [423]:
data['price'].head()

0    269 000        PLN
1     29 900        PLN
2     60 000        PLN
3     75 000        PLN
4     75 000        PLN
Name: price, dtype: object

In [424]:
# There are some prices with "PLN" and some with "EUR", so for this column different approach is needed

# Extracting the value and currency from the price column
data['value'] = data['price'].map(lambda x: float(''.join(x.split()[:-1]).replace(',', '.')))
data['currency'] = data['price'].map(lambda x: x.split()[-1])

data.head()

brand model    year   mileage  engine_capacity fuel_type  power  \
0  Audi   SQ7  2017.0  142800.0           3956.0    Diesel  435.0   
1  Audi    A3  2010.0  236000.0           1395.0   Benzyna  125.0   
2  Audi    A4  2018.0   84500.0           1968.0    Diesel  190.0   
3  Audi    Q5  2018.0   38000.0           1984.0   Benzyna  252.0   
4  Audi    A7  2016.0   74885.0           2995.0   Benzyna  333.0   

   transmission             drive  city_fuel_consumption  ... doors  seats  \
0  Automatyczna       4x4 (stały)                    8.0  ...   5.0    7.0   
1      Manualna  Na przednie koła                    NaN  ...   5.0    5.0   
2  Automatyczna       4x4 (stały)                    5.0  ...   5.0    5.0   
3  Automatyczna       4x4 (stały)                    8.0  ...   5.0    5.0   
4  Automatyczna       4x4 (stały)                   10.0  ...   5.0    5.0   

    color origin_country accident_free serviced_at_ASO condition  \
0   Szary         Polska           Tak             Tak   Używane   
1  Czarny            NaN           Tak             Tak   Używane   
2   Szary            NaN           NaN             NaN   Używane   
3   Szary            NaN           NaN             NaN   Używane   
4   Biały            NaN           NaN             NaN   Używane   

                price     value  currency  
0  269 000        PLN  269000.0       PLN  
1   29 900        PLN   29900.0       PLN  
2   60 000        PLN   60000.0       PLN  
3   75 000        PLN   75000.0       PLN  
4   75 000        PLN   75000.0       PLN  

[5 rows x 21 columns]

In [425]:
# Check the rows with EUR currency
eur_price = data[data['currency'] == 'EUR']

print(f'There are {len(eur_price)} rows with EUR currency.')
eur_price.head()

There are 224 rows with EUR currency.


brand         model    year   mileage  engine_capacity    fuel_type  \
1416  Audi  RS e-tron GT  2023.0      10.0              NaN  Elektryczny   
1431  Audi  RS e-tron GT  2023.0      15.0              NaN  Elektryczny   
3770  Audi            A8  2018.0  114699.0           2967.0       Diesel   
6765  Audi            A4  2020.0   22000.0           1968.0       Diesel   
6786  Audi           RS3  2022.0    6500.0           2480.0      Benzyna   

      power  transmission        drive  city_fuel_consumption  ... doors  \
1416  598.0  Automatyczna  4x4 (stały)                    NaN  ...   4.0   
1431  598.0  Automatyczna  4x4 (stały)                    NaN  ...   4.0   
3770  286.0  Automatyczna  4x4 (stały)                    8.0  ...   4.0   
6765  190.0  Automatyczna  4x4 (stały)                    5.0  ...   5.0   
6786  400.0  Automatyczna          NaN                   11.1  ...   4.0   

      seats       color origin_country accident_free serviced_at_ASO  \
1416    NaN      Czarny            NaN           NaN             NaN   
1431    NaN     Zielony            NaN           NaN             NaN   
3770    5.0      Czarny         Niemcy           NaN             NaN   
6765    5.0  Inny kolor            NaN           NaN             Tak   
6786    5.0       Szary            NaN           NaN             NaN   

     condition              price    value  currency  
1416   Używane  69 000        EUR  69000.0       EUR  
1431   Używane  70 000        EUR  70000.0       EUR  
3770   Używane  47 000        EUR  47000.0       EUR  
6765   Używane  10 500        EUR  10500.0       EUR  
6786   Używane  55 000        EUR  55000.0       EUR  

[5 rows x 21 columns]

In [426]:
# Convert EUR to PLN
data.loc[data['currency'] == 'EUR', 'value'] *= 4.45

In [427]:
# Drop no longer needed columns
data = data.drop(['price', 'currency'], axis='columns')

data = data.rename(columns={"value": "price"})

data.head()

brand model    year   mileage  engine_capacity fuel_type  power  \
0  Audi   SQ7  2017.0  142800.0           3956.0    Diesel  435.0   
1  Audi    A3  2010.0  236000.0           1395.0   Benzyna  125.0   
2  Audi    A4  2018.0   84500.0           1968.0    Diesel  190.0   
3  Audi    Q5  2018.0   38000.0           1984.0   Benzyna  252.0   
4  Audi    A7  2016.0   74885.0           2995.0   Benzyna  333.0   

   transmission             drive  city_fuel_consumption body_type  doors  \
0  Automatyczna       4x4 (stały)                    8.0       SUV    5.0   
1      Manualna  Na przednie koła                    NaN   Kompakt    5.0   
2  Automatyczna       4x4 (stały)                    5.0     Kombi    5.0   
3  Automatyczna       4x4 (stały)                    8.0       SUV    5.0   
4  Automatyczna       4x4 (stały)                   10.0   Kompakt    5.0   

   seats   color origin_country accident_free serviced_at_ASO condition  \
0    7.0   Szary         Polska           Tak             Tak   Używane   
1    5.0  Czarny            NaN           Tak             Tak   Używane   
2    5.0   Szary            NaN           NaN             NaN   Używane   
3    5.0   Szary            NaN           NaN             NaN   Używane   
4    5.0   Biały            NaN           NaN             NaN   Używane   

      price  
0  269000.0  
1   29900.0  
2   60000.0  
3   75000.0  
4   75000.0

## **Categorical columns**